src/libraries/System.Collections/src/System/Collections/Generic/PriorityQueue.cs

```diff
@@ -0,0 +1,695 @@
+        /// Sets the capacity to the actual number of items in the priority queue,
+        /// if that is less than 90 percent of current capacity.
+        /// </summary>
+        public void TrimExcess()
+        {
+            int threshold = (int)(((double)_nodes.Length) * 0.9);
+            if (_size < threshold)
+            {
+                Array.Resize(ref _nodes, _size);
+                _version++;
+            }
+        }
+
+        private void EnsureEnoughCapacityBeforeAddingNode()
+        {
+            if (_size == _nodes.Length)
+            {
+                int newCapacity = (int)((long)_nodes.Length * (long)GrowFactor / 100);
+                if (newCapacity < _nodes.Length + MinimumGrow)
+                {
+                    newCapacity = _nodes.Length + MinimumGrow;
+                }
+                SetCapacity(newCapacity);
+            }
+        }
+
+        /// <summary>
+        /// Grows or shrinks the array holding nodes. Capacity must be >= _size.
+        /// </summary>
+        private void SetCapacity(int capacity)
+        {
+            var newArray = new (TElement, TPriority)[capacity];
+
+            if (_size > 0)
+            {
+                Array.Copy(_nodes, 0, newArray, 0, _size);
+            }
+
+            _nodes = newArray;
+        }
+
+        /// <summary>
+        /// Removes the node at the specified index.
+        /// </summary>
+        private void Remove(int indexOfNodeToRemove)
+        {
+            // The idea is to replace the specified node by the very last
+            // node and shorten the array by one.
+
+            int lastNodeIndex = GetLastNodeIndex();
+            var lastNode = _nodes[lastNodeIndex];
+            _size--;
+            _version++;
+
+            // In case we wanted to remove the node that was the last one,
+            // we are done.
+
+            if (indexOfNodeToRemove == lastNodeIndex)
+            {
+                return;
+            }
+
+            // Our last node was erased from the array and needs to be
+            // inserted again. Of course, we will overwrite the node we
+            // wanted to remove. After that operation, we will need
+            // to restore the heap property (in general).
+
+            var nodeToRemove = _nodes[indexOfNodeToRemove];
+
+            int relation = this.Comparer.Compare(lastNode.priority, nodeToRemove.priority);
+            PutAt(lastNode, indexOfNodeToRemove);
+
+            if (relation < 0)
+            {
+                MoveUp(lastNode, indexOfNodeToRemove);
+            }
+            else
+            {
+                MoveDown(lastNode, indexOfNodeToRemove);
+            }
+        }
+
+        /// <summary>
+        /// Puts a node at the specified index.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void PutAt((TElement element, TPriority priority) node, int index)
+        {
+            _nodes[index] = node;
+        }
```

Comment:
From my perspective, this helper hurts rather than helps clarity of the calling code.  Just as a method like `Remove` directly accesses `_nodes` to "GetAt", methods that use `PutAt` should directly do `_nodes[index] = node` rather than `PutAt(node, index)`.